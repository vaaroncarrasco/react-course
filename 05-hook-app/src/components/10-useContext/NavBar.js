import React from 'react'
import { NavLink } from 'react-router-dom';

// Link -> like a
// NavLink -> like a but with exact & activeClassName attribute

export const NavBar = () => {
    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">

            <NavLink to="/" className="navbar-brand">useContext</NavLink>

            <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
                <div className="navbar-nav">
                    <NavLink exact activeClassName="active" to="/" className="nav-item nav-link">Home</NavLink>
                    <NavLink exact activeClassName="active" to="/about" className="nav-item nav-link">About</NavLink>
                    <NavLink exact activeClassName="active" to="/login" className="nav-item nav-link">Login</NavLink>
                </div>

            </div>

        </nav>
    )
}
