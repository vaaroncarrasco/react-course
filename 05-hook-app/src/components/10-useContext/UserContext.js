import { createContext } from 'react';

// * Higher Order Component

// 1. createContext()
export const UserContext = createContext(null);