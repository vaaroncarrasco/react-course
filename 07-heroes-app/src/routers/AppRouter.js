import React, { useContext } from 'react'
import {
    BrowserRouter as Router,
    Switch
  } from 'react-router-dom';

import { AuthContext } from '../auth/AuthContext';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';


export const AppRouter = () => {

    // ? Only principal routers are wrapped in a <Router></Router> HOC component with a <Switch></Switch> component inside.
    // * Child-router is just a function component. It has only a <Switch></Switch> without <Router></Router>

    const { user } = useContext(AuthContext);

    return (
        <Router>
            <div>
                {/* We can set global style or html elements in here. They'd appear in every component */}
                <Switch>
                    <PublicRoute
                        exact
                        path="/login"
                        component={ LoginScreen }
                        isAuthenticated={ user.logged }
                    />

                    {/* All routes pass through this non-exact route "/" -> DashboardRoutes */}
                    <PrivateRoute
                        path="/"
                        component={ DashboardRoutes }
                        isAuthenticated={ user.logged }
                    />
                </Switch>
            </div>
        </Router>
    )
}
