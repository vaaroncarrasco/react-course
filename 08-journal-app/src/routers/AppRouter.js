import React, { useEffect, useState } from 'react';
import {
    BrowserRouter as Router,
    Switch,
    Redirect
  } from 'react-router-dom';

import { AuthRouter } from './AuthRouter';
import { JournalScreen } from '../components/journal/JournalScreen';
import { firebase } from '../firebase/firebase-config';
import { useDispatch } from 'react-redux';
import { login } from '../actions/auth';
import { PublicRoute } from './PublicRoute';
import { PrivateRoute } from './PrivateRoute';
import { startLoadingNotes } from '../actions/notes';

export const AppRouter = () => {

    const dispatch = useDispatch();

    // * Check if user is authenticated
    const [checking, setChecking] = useState(true);

    const [isLoggedIn, setIsLoggedIn] = useState(false);

    //? Maintain state using firebase
    useEffect(() => {

        // * onAuthStateChanged -> creates an Observable
        // Observable -> special obj that is triggered depending on events

        // ? Async cause loadNotes() returns promise
        firebase.auth().onAuthStateChanged( async (user) => {

            if ( user?.uid ) {
                dispatch( login(user.uid, user.displayName) );

                setIsLoggedIn( true );

                // * Components should only dispatch and actions do heavy tasks
                dispatch( startLoadingNotes( user.uid ) )

            } else {
                setIsLoggedIn( false );
            }

            setChecking(false);

        })

    }, [dispatch, setChecking, setIsLoggedIn]); // only runs once

    if (checking) {
        return (
            <h1>Wait...</h1>
        )
    }

    return (
        <Router>
            <div>
                <Switch>
                    <PublicRoute
                        isAuthenticated={ isLoggedIn }
                        path="/auth"
                        component={ AuthRouter }
                    />

                    <PrivateRoute
                        isAuthenticated={ isLoggedIn }
                        exact
                        path="/"
                        component={ JournalScreen }
                    />

                    <Redirect to="/auth/login" />

                </Switch>
            </div>
        </Router>
    )
}
