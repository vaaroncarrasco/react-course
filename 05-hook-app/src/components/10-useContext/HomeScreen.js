import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const HomeScreen = () => {

    // 3. useContext -> use createdContext to get data -> import context component and useContext()
    const { user } = useContext(UserContext);

    return (
        <div>
            <h1>HomeScreen</h1>
            <hr />

            <pre>
                {
                    JSON.stringify( user, null, 3)
                }
            </pre>
        </div>
    )
}
