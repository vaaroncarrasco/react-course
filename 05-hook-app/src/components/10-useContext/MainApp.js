import React, { useState } from 'react'
import { AppRouter } from './AppRouter'
import { UserContext } from './UserContext'

export const MainApp = () => {

    const [user, setUser] = useState({});

    return (
        // * Higher Order Component creates -> Context.Provider -> allowing to pass data thru child components
        // 2. Send data thru context component.Provider value={ value }
        <UserContext.Provider value={{
            user,
            setUser
        }}>

            <AppRouter />

        </UserContext.Provider>
    )
}
