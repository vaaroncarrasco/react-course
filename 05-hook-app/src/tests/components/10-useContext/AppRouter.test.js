import { mount } from 'enzyme';
import React from 'react';
import { AppRouter } from '../../../components/10-useContext/AppRouter';
import { UserContext } from '../../../components/10-useContext/UserContext';

describe('testing <AppRouter />', () => {

    const user = {
        id: 1,
        name: 'Fernando'
    }

    const wrapper = mount(
        <UserContext.Provider value={ {user} }>
            <AppRouter />
        </UserContext.Provider>
    );

    test('should match snapshot', () => {

        expect( wrapper ).toMatchSnapshot();
    });

});
