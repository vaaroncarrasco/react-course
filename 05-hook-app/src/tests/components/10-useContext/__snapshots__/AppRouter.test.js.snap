// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <AppRouter /> should match snapshot 1`] = `
<div>
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <a
      aria-current="page"
      className="navbar-brand active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      useContext
    </a>
    <div
      className="collapse navbar-collapse"
      id="navbarNavAltMarkup"
    >
      <div
        className="navbar-nav"
      >
        <a
          aria-current="page"
          className="nav-item nav-link active"
          href="/"
          onClick={[Function]}
          style={Object {}}
        >
          Home
        </a>
        <a
          aria-current={null}
          className="nav-item nav-link"
          href="/about"
          onClick={[Function]}
        >
          About
        </a>
        <a
          aria-current={null}
          className="nav-item nav-link"
          href="/login"
          onClick={[Function]}
        >
          Login
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <div>
      <h1>
        HomeScreen
      </h1>
      <hr />
      <pre>
        {
   "id": 1,
   "name": "Fernando"
}
      </pre>
    </div>
  </div>
</div>
`;
